### PR TITLE
[fix] Combination of autocreate + forced delete of partitioned topic with active consumer leaves topic metadata inconsistent

### PR DESCRIPTION
**Pull request by dlg99**
https://github.com/apache/pulsar/pull/17308 -> https://github.com/apache/pulsar/pull/17566 -> this

disables autocreation per topic. skips system topics.

testing locally

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/service/PersistentTopicTest.java [hunk 1]
```diff
@@ -1310,12 +1310,23 @@ public void testDeleteTopic() throws Exception {
...
         f1.get();
 
-        assertTrue(topic.delete().isCompletedExceptionally());
+        assertTrue(waitForDone(topic.delete()).isCompletedExceptionally());
         assertFalse((boolean) isFencedField.get(topic));
         assertFalse((boolean) isClosingOrDeletingField.get(topic));
         topic.unsubscribe(successSubName);
     }
 
+    private CompletableFuture<Void> waitForDone(CompletableFuture<Void> cf) throws InterruptedException {
```
Comment (eolivelli) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/service/PersistentTopicTest.java:1319]: you can use Awaiatility
`await().until(() -> cf.isDone())`

**Review by eolivelli — COMMENTED** [review]
I think that a better approach is to add a "deleted" field to PartitionedTopicMetadata
https://github.com/apache/pulsar/blob/996b30187686722ede254d9f11723d62384e4909/pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/partition/PartitionedTopicMetadata.java#L33


```
public class PartitionedTopicMetadata {
    public int partitions;
    public Map<String, String> properties;
    public boolean deleted;
    }
```
    
when we start the deletion of the partitioned topic we set it to "deleted=true"
when something wants to auto-create the partitions or to access the topic we prevent that action.

this would be consistent with the way we are doing it for the Namespace:
https://github.com/apache/pulsar/blob/6c7fdf204e07b8fc8d589f813d71ef201f6c0e3e/pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/Policies.java#L66

similar PR
https://github.com/apache/pulsar/pull/17609

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java [hunk 2]
```diff
@@ -295,6 +296,75 @@ public CompletableFuture<Void> clearPartitionedTopicTenantAsync(String tenant) {
             final String partitionedTopicPath = joinPath(PARTITIONED_TOPIC_PATH, tenant);
             return deleteIfExistsAsync(partitionedTopicPath);
         }
+
+        public CompletableFuture<Void> markPartitionedTopicDeletedAsync(TopicName tn) {
+            if (tn.isPartitioned()) {
+                return CompletableFuture.completedFuture(null);
+            }
+            log.debug("markPartitionedTopicDeletedAsync {}", tn);
```
Comment (eolivelli) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java:304]: nit: across the code base we use "if log.isDebugEnabled" guards.
we are not in a hot path but that's the style used by the project

**Review by eolivelli — APPROVED** [review]
LGTM

Great work!

(I have left a minor comment about logger)

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java (hunk 2 above)
Comment (dlg99) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java:304]: I don't think it is needed there. "if log.isDebugEnabled" guards against unnecessary boxing of primitives, tn is an Object already.

**Pull request closed without merging**